src/detachment_controller.cxx

```diff
@@ -0,0 +1,229 @@
+#include "../include/detachment_controller.hxx"
+
+#include <bout/mesh.hxx>
+using bout::globals::mesh;
+#include <algorithm>  // Include for std::max
+#include <iostream>
+#include <iomanip>
+#include <numeric>
+
+BoutReal calculateGradient(const std::vector<BoutReal>& x, const std::vector<BoutReal>& y) {
+    size_t N = x.size();
+    BoutReal sum_x = std::accumulate(x.begin(), x.end(), 0.0);
+    BoutReal sum_y = std::accumulate(y.begin(), y.end(), 0.0);
+    BoutReal sum_xy = 0.0;
+    BoutReal sum_x_squared = 0.0;
+    
+    for (size_t i = 0; i < N; ++i) {
+        sum_xy += x[i] * y[i];
+        sum_x_squared += x[i] * x[i];
+    }
+    
+    BoutReal numerator = N * sum_xy - sum_x * sum_y;
+    BoutReal denominator = N * sum_x_squared - sum_x * sum_x;
+    
+    if (denominator == 0) {
+        return 0.0; // Handle division by zero or return an error code
+    }
+
+    return numerator / denominator;
+}
+
+void DetachmentController::transform(Options& state) {
+
+    std::cout << std::fixed << std::setprecision(15);
+
+    // Part 1: compute the detachment front location
+    Field3D neutral_density = getNoBoundary<Field3D>(state["species"][neutral_species]["density"]);
+    Field3D electron_density = getNoBoundary<Field3D>(state["species"]["e"]["density"]);
+    Coordinates *coord = mesh->getCoordinates();
+
+    // Set the initial value so that if no point has Nn > Ne, the detachment front is
+    // at the target.
+
+    bool detachment_front_found = false;
+    BoutReal distance_from_upstream = -(coord->dy(0, mesh->ystart - 1, 0) + coord->dy(0, mesh->ystart, 0)) / 4.0;
```

Comment:
To work in parallel, this distance from upstream should be calculated once in the constructor